packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -3227,14 +3233,22 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
   }
 
   @override
-  InheritedWidget inheritFromWidgetOfExactType(Type targetType) {
+  InheritedWidget inheritFromWidgetOfExactType(Type targetType, { dynamic scope }) {
     assert(_debugCheckStateIsActiveForAncestorLookup());
     final InheritedElement ancestor = _inheritedWidgets == null ? null : _inheritedWidgets[targetType];
     if (ancestor != null) {
       assert(ancestor is InheritedElement);
       _dependencies ??= new HashSet<InheritedElement>();
       _dependencies.add(ancestor);
-      ancestor._dependents.add(this);
+      // If we're ever called with scope == null (the common case) then
+      // InheritedElement.notifyClients will unconditionally rebuild all dependents.
```

Comment:
This comment seems like it doesn't accurately describe what's going on here, in a way that actually quite dramatically changes my opinion of the patch. :-)

I think what this code is actually doing is better described by:
```
// If the `scope` is null, then always update this build context when
// the inherited widget changes, regardless of which scopes changed.
```

But then you also need to distinguish "we were once called with null" from "we have never been called" and from "we were once called with null then called with a scope" and "we were once called with a scope then called with null" (the last two should be the same).

(Side note: the documentation needs updating too.)